File: golang/maximal-rectangle/maximal-rectangle.go
package main


import (
	"fmt"
)

func maximalRectangle(matrix [][]byte) int {
	if len(matrix) == 0 {
		return 0
	}
	cols := len(matrix[0])
	heights := make([]int, cols)
	maxArea := 0

	for _, row := range matrix {
		for j, val := range row {
			if val == '1' {
				heights[j]++
			} else {
				heights[j] = 0
			}
		}
		currentArea := largestRectangleArea(heights)
		if currentArea > maxArea {
			maxArea = currentArea
		}
	}
	return maxArea
}

func largestRectangleArea(heights []int) int {
	stack := []int{-1}
	maxArea := 0
	for i := 0; i <= len(heights); i++ {
		currentH := 0
		if i < len(heights) {
			currentH = heights[i]
		}
		for len(stack) > 1 && currentH < heights[stack[len(stack)-1]] {
			h := heights[stack[len(stack)-1]]
			stack = stack[:len(stack)-1]
			w := i - stack[len(stack)-1] - 1
			if h*w > maxArea {
				maxArea = h*w
			}
		}
		stack = append(stack, i)
	}
	return maxArea
}

func main() {
	makeMatrix := func(rows []string) [][]byte {
		mat := make([][]byte, len(rows))
		for i, s := range rows {
			mat[i] = []byte(s)
		}
		return mat
	}

	runTest := func(testNum int, input []string, expected int) {
		matrix := makeMatrix(input)
		result := maximalRectangle(matrix)
		status := "FAIL"
		if result == expected {
			status = "PASS"
		}
		fmt.Printf("Test Case %d: Expected %d, Got %d -> %s\n", testNum, expected, result, status)
	}

	input1 := []string{
		"10100",
		"10111",
		"11111",
		"10010",
	}
	runTest(1, input1, 6)

	input2 := []string{"0"}
	runTest(2, input2, 0)

	input3 := []string{"1"}
	runTest(3, input3, 1)

	input4 := []string{
		"11",
		"11",
	}
	runTest(4, input4, 4)
}

File: leetcode/maximal-rectangle.py
class Solution(object):
    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        
        if not matrix:
            return 0
            
        cols = len(matrix[0])
        heights = [0] * cols
        max_area = 0
        
        for row in matrix:
            for i, val in enumerate(row):
                # Update histogram heights
                if val == '1':
                    heights[i] += 1
                else:
                    heights[i] = 0
            

    
            # Calculate max area for the current row's histogram
            max_area = max(max_area, self.largestRectangleArea(heights))
            
        return max_area

    def largestRectangleArea(self, heights):
        # Add a 0 at the end to force the stack to empty
        stack = [-1]
        max_area = 0


        for i in range(len(heights) + 1):
            current_h = heights[i] if i < len(heights) else 0
            
            while len(stack) > 1 and current_h < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i - stack[-1] - 1
                max_area = max(max_area, h * w)
            
            stack.append(i)
            
        return max_area

if __name__ == "__main__":
    solver = Solution()
    
    def run_test(case_num, matrix, expected):
        result = solver.maximalRectangle(matrix)
        status = "PASS" if result == expected else "FAIL"
        print(f"Test Case {case_num}: {result} (Expected: {expected}) -> {status}")

    # Example 1
    matrix1 = [
        ["1","0","1","0","0"],
        ["1","0","1","1","1"],
        ["1","1","1","1","1"],
        ["1","0","0","1","0"]
    ]
    run_test(1, matrix1, 6)

    # Example 2
    matrix2 = [["0"]]
    run_test(2, matrix2, 0)

    # Example 3
    matrix3 = [["1"]]
    run_test(3, matrix3, 1)

    # Example 4: Full square
    matrix4 = [
        ["1","1"],
        ["1","1"]
    ]
    run_test(4, matrix4, 4)